p4-16/psa/PSA.mdk

```diff
@@ -561,11 +598,67 @@ extern Counter<W, S> {
 }
 ```
 
+See below for pseudocode of a reference implementation for the Counter
+extern.
+
+```
+Counter(N n_counters) {
+    this.num_counters = n_counters;
+    this.counter_vals = new array of size n_counters, each element with type W;
+    this.max_counter_value = maximum value possible for type W;
+}
+
+/// Add @add_value to a saturating counter whose current value is
+/// @cur_value, and whose maximum possible value is @max_value.
+///
+/// Assume that type W is bit<X> and type V is bit<Y> where X >= Y.
+/// There is very little, if any, reason to use signed count values.
+
+W next_counter_value(W cur_value, V add_value, W max_value) {
+    bit<X+1> next_value = (bit<X+1>) cur_value + (bit<X+1>) add_value;
+    if (next_value[X] == 1) {
+        return max_value;
+    }
+    return (bit<X>) next_value;
+}
+
+void count(in S index, in V add_value) {
+    if (index < this.num_counters) {
+        this.counter_vals[index] = next_counter_value(this.counter_vals[index],
+                                           add_value, this.max_counter_value);
+    } else {
+        // No counter_vals updated if index is out of range.
+        // See below for optional debug information to record.
+    }
+}
+
+void count(in S index) {
+    count(index, 1);
+}
+```
+
+Rationale for saturating behavior: With saturating behavior, the
+control plane can know that information was lost, if it reads the
+maximum possible value from a counter.  In this case, it knows that
```

Comment:
@vgurevich actually we can control the programmer's behavior to a certain extent by making it harder to code in a bad way. Giving them safe default behavior and the threads to build the rope to hang themselves will always push them toward the safe behavior because it's too much work to twist enough rope.